plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Inline.hs

```diff
@@ -419,9 +482,15 @@ effectSafe
   -> Bool
   -- ^ is it pure?
   -> InlineM name uni fun a Bool
-effectSafe body n purity = do
-  immediatelyEvaluated <- isFirstVarBeforeEffects n body
-  pure $ purity || immediatelyEvaluated
+effectSafe body n termIsPure = do
+  preserveLogging <- view iiPreserveLogging
+  builtinSemantics <- view iiBuiltinSemanticsVariant
+  return $
+    termIsPure
+      || isFirstVarBeforeEffects builtinSemantics n body
+      || (not preserveLogging && isVarImmediatelyEvaluated n body)
+ where
+  isVarImmediatelyEvaluated name = maybe False not . isVarDelayed name
```

Comment:
`isVarImmediatelyEvaluated` -> `isVarEventuallyEvaluated` (or just `eventuallyEvaluated` since it is too long) since the var does not have to be the first thing that gets evaluated